### PR TITLE
docs: added OpenAPI spec for rbac-backend

### PR DESCRIPTION
**Pull request by 04kash**
Added OpenAPI spec for `rbac-backend`
Fixes Issue: https://issues.redhat.com/browse/RHIDP-2761

### CODE REVIEW

**Comment by 04kash** [comment]
> Is there a process/script/automation for generating the OpenAPI spec when changes are made to the plugin?

I haven't set up anything for this yet, I can look into it.

File: plugins/rbac-backend/openapi.json
Comment (rm3l) [inline plugins/rbac-backend/openapi.json:1]: Would it make sense to convert this to YAML instead? More user-friendly to me :-)

**Comment by AndrienkoAleksandr** [comment]
Can you share, please, some docs, instruction, how we can update it manually to support this file in actual state?

**Comment by AndrienkoAleksandr** [comment]
> Could you please clarify your question? Do you mean how changes to the plugin code will be reflected in the OpenAPI spec?

Yes, and I thought this spec was generated with the help of some analyzer tool. Let me know if I'm wrong and it was done manually

**Comment by 04kash** [comment]
> > Could you please clarify your question? Do you mean how changes to the plugin code will be reflected in the OpenAPI spec?
> 
> Yes, and I thought this spec was generated with the help of some cod analyzer tool. Let me know if I'm wrong and it was done manually

The spec was made manually. So when changes are made to the plugin code, the OpenAPI spec will have to be updated manually.

File: plugins/rbac-backend/openapi.yaml [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+      description: Lists all roles
+      responses:
+        '200':
+          description: Request was successful.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  '#ref': '#/components/schemas/RoleResponse'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:137]:
```suggestion
                  '$ref': '#/components/schemas/RoleResponse'
```

File: plugins/rbac-backend/openapi.yaml [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+        '403':
+          description: Refusal to authorize
+    post:
+      description: Creates a new role.
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              '#ref': '#/components/schemas/Role'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:147]:
```suggestion
              '$ref': '#/components/schemas/Role'
```

File: plugins/rbac-backend/openapi.yaml [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+        - $ref: '#/components/parameters/nameParam'
+        - $ref: '#/components/parameters/namespaceParam'
+        - $ref: '#/components/parameters/kindParam'
+      responses:
+        '200':
+          description: Request was successful.
+          content:
+            application/json:
+              schema:
+                '#ref': '#/components/schemas/RoleResponse'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:170]:
```suggestion
                '$ref': '#/components/schemas/RoleResponse'
```

File: plugins/rbac-backend/openapi.yaml [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+        - $ref: '#/components/parameters/kindParam'
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              type: object
+              properties:
+                oldRole:
+                  '#ref': '#/components/schemas/Role'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:189]:
```suggestion
                  '$ref': '#/components/schemas/Role'
```

File: plugins/rbac-backend/openapi.yaml [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+                newRole:
+                  '#ref': '#/components/schemas/Role'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:191]:
```suggestion
                  '$ref': '#/components/schemas/Role'
```

File: plugins/rbac-backend/openapi.yaml [hunk 6]
```diff
@@ -0,0 +1,636 @@
...
+      description: Lists all permission polices.
+      responses:
+        '200':
+          description: Request was successful.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  '#ref': '#/components/schemas/PermissionResponse'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:231]:
```suggestion
                  '$ref': '#/components/schemas/PermissionResponse'
```

File: plugins/rbac-backend/openapi.yaml [hunk 7]
```diff
@@ -0,0 +1,636 @@
...
+    post:
+      description: Creates one or more permission policies for a specified entity.
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              type: array
+              items:
+                '#ref': '#/components/schemas/PermissionPolicy'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:243]:
```suggestion
                '$ref': '#/components/schemas/PermissionPolicy'
```

File: plugins/rbac-backend/openapi.yaml [hunk 8]
```diff
@@ -0,0 +1,636 @@
...
+        - $ref: '#/components/parameters/kindParam'
+      responses:
+        '200':
+          description: Request was successful.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  '#ref': '#/components/schemas/PermissionResponse'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:266]:
```suggestion
                  '$ref': '#/components/schemas/PermissionResponse'
```

File: plugins/rbac-backend/openapi.yaml [hunk 9]
```diff
@@ -0,0 +1,636 @@
...
+          schema:
+            type: string
+      requestBody:
+        required: false
+        content:
+          application/json:
+            schema:
+              type: array
+              items:
+                '#ref': '#/components/schemas/PermissionPolicy'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:360]:
```suggestion
                '$ref': '#/components/schemas/PermissionPolicy'
```

File: plugins/rbac-backend/openapi.yaml [hunk 10]
```diff
@@ -0,0 +1,636 @@
+openapi: 3.1.0
```
[the remark this answers is not in the document]
Comment (AndrienkoAleksandr) [reply plugins/rbac-backend/openapi.yaml:1]: @PatAKnight do you agree with downgrade version?

File: plugins/rbac-backend/openapi.yaml [hunk 11]
```diff
@@ -0,0 +1,636 @@
...
+          metadata:
+            type: object
+            description: Metadata about the role.
+            properties:
+              source:
+                type: string
+                description: The source from which the role was defined.
+              description:
+                type: string
+                description: A description of the role.
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:35]:
```suggestion
              author:
                type: string
                description: The author of the role.
              createdAt:
                type: string
                description: The date and time the role was created.
              lastModified:
                type: string
                description: The date and time the role was last modified.
              modifiedBy:
                type: string
                description: The user who last modified the role.
              source:
                type: string
                description: The source from which the role was defined.
              description:
                type: string
                description: A description of the role.```

File: plugins/rbac-backend/openapi.yaml [hunk 12]
```diff
@@ -0,0 +1,636 @@
...
+          type: string
+          description: The name of the role.
+        metadata:
+          type: object
+          description: Metadata about the role.
+          properties:
+            description:
+              type: string
+              description: A description of the role.
+
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:54]:
```suggestion
    Condition:
      type: object
      oneOf:
        - properties:
            anyOf:
              type: array
              items:
                $ref: '#/components/schemas/Condition'
          required: [anyOf]
        - properties:
            allOf:
              type: array
              items:
                $ref: '#/components/schemas/Condition'
          required: [allOf]
        - properties:
            not:
              $ref: '#/components/schemas/Condition'
          required: [not]
        - properties:
            rule:
              type: string
            resourceType:
              type: string
            params:
              type: object
          required: [rule, resourceType, params]
    PropertyObject:
      type: object
      properties:
        type:
          type: string
        description:
          type: string
      required: [type, description]
    PropertyArray:
        type: object
        properties:
          type:
            type: string
          description:
            type: string
          items:
            type: object
            properties:
              type:
                type: string
        required: [type, description, items]

```

File: plugins/rbac-backend/openapi.yaml [hunk 13]
```diff
@@ -0,0 +1,636 @@
...
+      description: Lists all roles
+      responses:
+        '200':
+          description: Request was successful.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  '$ref': '#/components/schemas/RoleResponse'
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:137]:
```suggestion
                type: object
                '$ref': '#/components/schemas/RoleResponse'
```

File: plugins/rbac-backend/openapi.yaml [hunk 14]
```diff
@@ -0,0 +1,636 @@
...
+      responses:
+        '204':
+          description: ok
+        '403':
+          description: Refusal to authorize
+        '404':
+          description: Could not find resource.
+  /api/permission/policies:
+    get:
+      description: Lists all permission polices.
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:222]:
```suggestion
      description: Lists all permission polices.
      parameters:
        - name: entityReference
          in: query
          description: Entity <kind>:<default>/<name>.
          required: false
          schema:
            type: string
        - name: permission
          in: query
          description: Permission from a specific plugin, Resource type or name
          required: false
          schema:
            type: string
        - name: policy
          in: query
          description: 'Policy action for the permission: create, read, update, delete, use'
          required: false
          schema:
            type: string
        - name: effect
          in: query
          description: allow or deny
          required: false
          schema:
            type: string
```

File: plugins/rbac-backend/openapi.yaml [hunk 15]
```diff
@@ -0,0 +1,636 @@
...
+                        isResourced:
+                          type: boolean
+                        permission:
+                          type: string
+                          description: Permission from a specific plugin.
+                        policy:
+                          type: string
+                          description: >-
+                            Policy action for the permission: create, read,
+                            update, delete, use.
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:397]: We have API changes with the latest main branch:
```suggestion
              schema:
                type: array
                items: 
                  type: object
                  properties:
                    pluginId:
                      type: string
                    policies:
                      type: array
                      items:
                        type: object
                        properties:
                          name:
                            type: string
                            description: Permission from a specific plugin.
                          resourceType:
                            type: string
                            description: Resource type.
                          policy:
                            type: string
                            description: >-
                              Policy action for the permission: create, read,
                              update, delete, use.
                        required: [name, policy]
```

File: plugins/rbac-backend/openapi.yaml [hunk 16]
```diff
@@ -0,0 +1,636 @@
...
+      description: Lists all conditions
+      responses:
+        '200':
+          description: Request was successful
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  type: object
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:449]:
```suggestion
                  type: object
                  required: [result, roleEntityRef, pluginId, resourceType, permissionMapping, conditions]
                  properties:
                    id:
                      type: integer
                    result:
                      type: string
                    roleEntityRef:
                      type: string
                    pluginId:
                      type: string
                    resourceType:
                      type: string
                    permissionMapping:
                      type: array
                      items:
                        type: string
                    conditions:
                      $ref: '#/components/schemas/Condition'
```

File: plugins/rbac-backend/openapi.yaml [hunk 17]
```diff
@@ -0,0 +1,636 @@
...
+                            type: array
+                            items:
+                              type: object
+                              properties:
+                                rule:
+                                  type: string
+                                resourceType:
+                                  type: string
+                                params:
+                                  type: object
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:480]:
```suggestion
                  type: object
                  required: [result, roleEntityRef, pluginId, resourceType, permissionMapping, conditions]
                  properties:
                    id:
                      type: integer
                    result:
                      type: string
                    roleEntityRef:
                      type: string
                    pluginId:
                      type: string
                    resourceType:
                      type: string
                    permissionMapping:
                      type: array
                      items:
                        type: string
                    conditions:
                      $ref: '#/components/schemas/Condition'
```

File: plugins/rbac-backend/openapi.yaml [hunk 18]
```diff
@@ -0,0 +1,636 @@
...
+                    type: string
+                conditions:
+                  type: object
+                  properties:
+                    rule:
+                      type: string
+                    resourceType:
+                      type: string
+                    params:
+                      type: object
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:512]:
```suggestion
                  $ref: '#/components/schemas/Condition'
```

File: plugins/rbac-backend/openapi.yaml [hunk 19]
```diff
@@ -0,0 +1,636 @@
...
+                        type: array
+                        items:
+                          type: object
+                          properties:
+                            rule:
+                              type: string
+                            resourceType:
+                              type: string
+                            params:
+                              type: object
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:563]:
```suggestion
                    $ref: '#/components/schemas/Condition'
```

File: plugins/rbac-backend/openapi.yaml [hunk 20]
```diff
@@ -0,0 +1,636 @@
...
+                      type: array
+                      items:
+                        type: object
+                        properties:
+                          rule:
+                            type: string
+                          resourceType:
+                            type: string
+                          params:
+                            type: object
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:610]:
```suggestion
                  $ref: '#/components/schemas/Condition'
```

File: plugins/rbac-backend/openapi.yaml [hunk 21]
```diff
@@ -0,0 +1,636 @@
...
+          description: 'Policy action for the permission: create, read, update, delete, use'
+          required: false
+          schema:
+            type: string
+        - name: effect
+          in: query
+          description: allow or deny
+          required: false
+          schema:
+            type: string
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:352]: I don't see these query params for delete method:
```suggestion
```

File: plugins/rbac-backend/openapi.yaml [hunk 22]
```diff
@@ -0,0 +1,636 @@
...
+                          type: string
+                          description: Permission from a specific plugin.
+                        policy:
+                          type: string
+                          description: >-
+                            Policy action for the permission: create, read,
+                            update, delete, use.
+        '403':
+          description: Refusal to authorize
+  /api/plugins/condition-rules:
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:400]:
```suggestion
  /api/permission/plugins/condition-rules:
```

File: plugins/rbac-backend/openapi.yaml [hunk 23]
```diff
@@ -0,0 +1,636 @@
...
+        '200':
+          description: Request was successful
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  type: object
+                  items:
+                    type: object
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:451]:
```suggestion
                  required: [result, roleEntityRef, pluginId, resourceType, permissionMapping, conditions]
```

File: plugins/rbac-backend/openapi.yaml (hunk 17 above)
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:480]:
```suggestion
                  properties:
                    id:
                      type: integer
                    result:
                      type: string
                    roleEntityRef:
                      type: string
                    pluginId:
                      type: string
                    resourceType:
                      type: string
                    permissionMapping:
                      type: array
                      items:
                        type: string
                    conditions:
                      $ref: '#/components/schemas/Condition'
```

File: plugins/rbac-backend/openapi.yaml [hunk 24]
```diff
@@ -0,0 +1,636 @@
...
+                          paramsSchema:
+                            type: object
+                          required:
+                            type: array
+                            items:
+                              type: string
+                          additionalProperties:
+                            type: boolean
+                          $schema:
+                            type: string
```
Comment (AndrienkoAleksandr) [inline plugins/rbac-backend/openapi.yaml:435]:
```suggestion
                            properties:
                              $schema:
                                type: string
                              additionalProperties:
                                type: boolean
                              required: 
                                type: string
                              type:
                                type: string
                            oneOf:
                              - properties:
                                  properties:
                                    type: object
                                    additionalProperties:
                                      $ref: '#/components/schemas/PropertyArray'
                              - properties:
                                  properties:
                                    type: object
                                    additionalProperties:
                                      $ref: '#/components/schemas/PropertyObject'
```

**Comment by AndrienkoAleksandr** [comment]
Apologies for the lengthy code review. I needed to gather some knowledge in the Open API area.

I have proposed the final fixes from the last code review round. After these are addressed, I am fine with merging.

File: plugins/rbac-backend/openapi.yaml (hunk 21 above)
Comment (04kash) [reply plugins/rbac-backend/openapi.yaml:412]: I saw them listed here: https://github.com/janus-idp/backstage-plugins/blob/main/plugins/rbac-backend/docs/apis.md#delete-permission

Comment (AndrienkoAleksandr) [reply plugins/rbac-backend/openapi.yaml:436]: I see, but this information is outdated. Can you fix also corresponding docs?

File: plugins/rbac-backend/openapi.yaml (hunk 18 above)
Comment (AndrienkoAleksandr) [reply plugins/rbac-backend/openapi.yaml:619]: ?

**Comment by AndrienkoAleksandr** [comment]
@04kash , sorry, can you rebase your pr also to make ci green?

File: plugins/rbac-backend/openapi.yaml (hunk 18 above)
Comment (04kash) [reply plugins/rbac-backend/openapi.yaml]: Was having some issues committing my changes yesterday, fixed that now

**Review by AndrienkoAleksandr — APPROVED** [review]

**Pull request merged**